IAM roles do not seem to allow Spark to access S3 files without explicit credentials

Flintrock: so fast! I'm super excited. I typically start launching a cluster from home, take the bus to work, and if I'm lucky, my cluster will be running when I get there. You write some pretty python too!

One nice feature of spark-ec2 is that it can copy AWS credentials and is setup (almost, have to export some things) right away to access data directly from s3. Since the data I use is exclusively in s3, this is a huge help. 

I don't have an obvious solution. I have a functioning core-site.xml that allows me to use hadoop from the command line, although I had to symlink /etc/hadoop to ~/hadoop/etc/hadoop. But that configuration doesn't seem to be recognized once I'm running spark, i.e. sc.textFile("s3://somewhere/somefile").first() raises an error that there is no s3/s3n framework. I am just hacking around, and I don't really know my way around the guts of hadoop (there are two! core-site.xml files, which do I use?), so there is surely a better way. 

I'm willing to contribute code/PRs too if someone can point me in the right direction!

Well, humpf. I made the change I mentioned above. I can now start to launch, the IAM role gets attached, Spark and Hadoop are installed, I can even login, but the launch command gives this error:

```
(SK16)marcus@lappy:XAdIPlus>flintrock launch MDCTest
Requesting 2 spot instances at a max price of $0.5...
0 of 2 instances granted. Waiting...
0 of 2 instances granted. Waiting...
All 2 instances granted.
[54.85.166.198] SSH online.
[54.85.166.198] Configuring ephemeral storage...
[52.90.219.236] SSH online.
[52.90.219.236] Configuring ephemeral storage...
[54.85.166.198] Installing HDFS...
[52.90.219.236] Installing HDFS...
[52.90.219.236] Installing Spark...
[54.85.166.198] Installing Spark...
[Errno -5] No address associated with hostname
Do you want to terminate the 2 instances created by this operation? [Y/n]: n
  File "/home/marcus/.local/lib/python3.4/site-packages/flintrock/flintrock.py", line 804, in main
    cli(obj={})
  File "/home/marcus/.local/lib/python3.4/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/home/marcus/.local/lib/python3.4/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/marcus/.local/lib/python3.4/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/marcus/.local/lib/python3.4/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/marcus/.local/lib/python3.4/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/marcus/.local/lib/python3.4/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/marcus/.local/lib/python3.4/site-packages/flintrock/flintrock.py", line 296, in launch
    instance_initiated_shutdown_behavior=ec2_instance_initiated_shutdown_behavior)
  File "/home/marcus/.local/lib/python3.4/site-packages/flintrock/ec2.py", line 31, in wrapper
    res = func(*args, **kwargs)
  File "/home/marcus/.local/lib/python3.4/site-packages/flintrock/ec2.py", line 526, in launch
    identity_file=identity_file)
  File "/home/marcus/.local/lib/python3.4/site-packages/flintrock/core.py", line 405, in provision_cluster
    identity_file=identity_file)
  File "/home/marcus/.local/lib/python3.4/site-packages/flintrock/ssh.py", line 64, in get_ssh_client
    timeout=3)
  File "/home/marcus/.local/lib/python3.4/site-packages/paramiko/client.py", line 237, in connect
    for (family, socktype, proto, canonname, sockaddr) in socket.getaddrinfo(hostname, port, socket.AF_UNSPEC, socket.SOCK_STREAM):
  File "/usr/lib/python3.4/socket.py", line 533, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
[Errno -5] No address associated with hostname
```
If I login anyway, and try to launch pyspark, I get this:
```[ec2-user@ip-172-30-0-30 ~]$ ./spark/bin/pyspark 
Python 2.7.10 (default, Aug 11 2015, 23:39:10) 
[GCC 4.8.3 20140911 (Red Hat 4.8.3-9)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
16/03/08 17:08:16 INFO SparkContext: Running Spark version 1.6.0
16/03/08 17:08:17 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
16/03/08 17:08:17 ERROR SparkContext: Error initializing SparkContext.
java.net.UnknownHostException: ip-172-30-0-30: ip-172-30-0-30: Name or service not known
	at java.net.InetAddress.getLocalHost(InetAddress.java:1496)
	at org.apache.spark.util.Utils$.findLocalInetAddress(Utils.scala:788)
	at org.apache.spark.util.Utils$.org$apache$spark$util$Utils$$localIpAddress$lzycompute(Utils.scala:781)
	at org.apache.spark.util.Utils$.org$apache$spark$util$Utils$$localIpAddress(Utils.scala:781)
	at org.apache.spark.util.Utils$$anonfun$localHostName$1.apply(Utils.scala:838)
	at org.apache.spark.util.Utils$$anonfun$localHostName$1.apply(Utils.scala:838)
	at scala.Option.getOrElse(Option.scala:120)
	at org.apache.spark.util.Utils$.localHostName(Utils.scala:838)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:420)
	at org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:59)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:234)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)
	at py4j.Gateway.invoke(Gateway.java:214)
	at py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:79)
	at py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:68)
	at py4j.GatewayConnection.run(GatewayConnection.java:209)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.UnknownHostException: ip-172-30-0-30: Name or service not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:922)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1316)
	at java.net.InetAddress.getLocalHost(InetAddress.java:1492)
	... 20 more
16/03/08 17:08:17 INFO SparkContext: Successfully stopped SparkContext
Traceback (most recent call last):
  File "/home/ec2-user/spark/python/pyspark/shell.py", line 43, in <module>
    sc = SparkContext(pyFiles=add_files)
  File "/home/ec2-user/spark/python/pyspark/context.py", line 115, in __init__
    conf, jsc, profiler_cls)
  File "/home/ec2-user/spark/python/pyspark/context.py", line 172, in _do_init
    self._jsc = jsc or self._initialize_context(self._conf._jconf)
  File "/home/ec2-user/spark/python/pyspark/context.py", line 235, in _initialize_context
    return self._jvm.JavaSparkContext(jconf)
  File "/home/ec2-user/spark/python/lib/py4j-0.9-src.zip/py4j/java_gateway.py", line 1064, in __call__
  File "/home/ec2-user/spark/python/lib/py4j-0.9-src.zip/py4j/protocol.py", line 308, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.
: java.net.UnknownHostException: ip-172-30-0-30: ip-172-30-0-30: Name or service not known
	at java.net.InetAddress.getLocalHost(InetAddress.java:1496)
	at org.apache.spark.util.Utils$.findLocalInetAddress(Utils.scala:788)
	at org.apache.spark.util.Utils$.org$apache$spark$util$Utils$$localIpAddress$lzycompute(Utils.scala:781)
	at org.apache.spark.util.Utils$.org$apache$spark$util$Utils$$localIpAddress(Utils.scala:781)
	at org.apache.spark.util.Utils$$anonfun$localHostName$1.apply(Utils.scala:838)
	at org.apache.spark.util.Utils$$anonfun$localHostName$1.apply(Utils.scala:838)
	at scala.Option.getOrElse(Option.scala:120)
	at org.apache.spark.util.Utils$.localHostName(Utils.scala:838)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:420)
	at org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:59)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:234)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)
	at py4j.Gateway.invoke(Gateway.java:214)
	at py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:79)
	at py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:68)
	at py4j.GatewayConnection.run(GatewayConnection.java:209)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.UnknownHostException: ip-172-30-0-30: Name or service not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:922)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1316)
	at java.net.InetAddress.getLocalHost(InetAddress.java:1492)
	... 20 more
```
I'm stumped. I have little knowledge of networking.

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,6 +8,7 @@ install:
 script:
   - "py.test ./tests/test_static.py"
   - "py.test ./tests/test_flintrock.py"
+  - "py.test ./tests/test_core.py"
   - "pip install -r requirements/maintainer.pip"
   - "py.test ./tests/test_pyinstaller_packaging.py"
 addons:

--- a/README.md
+++ b/README.md
@@ -18,10 +18,10 @@ Here's a quick way to launch a cluster on EC2, assuming you already have an [AWS
 ```sh
 flintrock launch test-cluster \
     --num-slaves 1 \
-    --spark-version 2.0.2 \
+    --spark-version 2.1.0 \
     --ec2-key-name key_name \
     --ec2-identity-file /path/to/key.pem \
-    --ec2-ami ami-b73b63a0 \
+    --ec2-ami ami-9be6f38c \
     --ec2-user ec2-user
 ```
 
@@ -56,6 +56,26 @@ flintrock <subcommand> --help
 ```
 
 That's not all. Flintrock has a few more [features](#features) that you may find interesting.
+
+### Accessing data on S3
+
+We recommend you access data on S3 from your Flintrock cluster by following
+these two steps:
+
+1. Setup an [IAM Role](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html)
+   that grants access to S3 as desired. Reference this role when you launch
+   your cluster using the `--ec2-instance-profile-name` option (or its
+   equivalent in your `config.yaml` file).
+2. In your Spark code reference paths using the `s3a://` prefix. `s3a://` is
+   backwards compatible with `s3n://` and replaces both `s3n://` and `s3://`.
+   The Hadoop project [recommends using `s3a://`](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html#S3A)
+   since it is actively developed, supports larger files, and offers
+   better performance.
+
+With this approach you don't need to copy around your AWS credentials
+or pass them into your Spark programs. As long as the assigned IAM role
+allows it, Spark will be able to read and write data to S3 simply by
+referencing the appropriate path (e.g. `s3a://bucket/path/to/file`).
 
 
 ## Installation
@@ -196,7 +216,7 @@ provider: ec2
 
 services:
   spark:
-    version: 2.0.2
+    version: 2.1.0
 
 launch:
   num-slaves: 1
@@ -207,7 +227,7 @@ providers:
     identity-file: /path/to/.ssh/key.pem
     instance-type: m3.medium
     region: us-east-1
-    ami: ami-b73b63a0
+    ami: ami-9be6f38c
     user: ec2-user
 ```
 

--- a/README.md
+++ b/README.md
@@ -60,17 +60,21 @@ That's not all. Flintrock has a few more [features](#features) that you may find
 ### Accessing data on S3
 
 We recommend you access data on S3 from your Flintrock cluster by following
-these two steps:
+these steps:
 
 1. Setup an [IAM Role](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html)
    that grants access to S3 as desired. Reference this role when you launch
    your cluster using the `--ec2-instance-profile-name` option (or its
    equivalent in your `config.yaml` file).
-2. In your Spark code reference paths using the `s3a://` prefix. `s3a://` is
+2. Reference S3 paths in your Spark code using the `s3a://` prefix. `s3a://` is
    backwards compatible with `s3n://` and replaces both `s3n://` and `s3://`.
    The Hadoop project [recommends using `s3a://`](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html#S3A)
    since it is actively developed, supports larger files, and offers
    better performance.
+3. Make sure Flintrock is configured to use Hadoop/HDFS 2.7+. Earlier
+   versions of Hadoop do not have solid implementations of `s3a://`.
+   Flintrock's default is Hadoop 2.7.3, so you don't need to do anything
+   here if you're using a vanilla configuration.
 
 With this approach you don't need to copy around your AWS credentials
 or pass them into your Spark programs. As long as the assigned IAM role

--- a/flintrock/config.yaml.template
+++ b/flintrock/config.yaml.template
@@ -1,6 +1,6 @@
 services:
   spark:
-    version: 2.0.2
+    version: 2.1.0
     # git-commit: latest  # if not 'latest', provide a full commit SHA; e.g. d6dc12ef0146ae409834c78737c116050961f350
     # git-repository:  # optional; defaults to https://github.com/apache/spark
     # optional; defaults to download from from the official Spark S3 bucket
@@ -24,7 +24,7 @@ providers:
     instance-type: m3.medium
     region: us-east-1
     # availability-zone: <name>
-    ami: ami-b73b63a0   # Amazon Linux, us-east-1
+    ami: ami-9be6f38c   # Amazon Linux, us-east-1
     user: ec2-user
     # ami: ami-61bbf104   # CentOS 7, us-east-1
     # user: centos

--- a/flintrock/core.py
+++ b/flintrock/core.py
@@ -423,30 +423,65 @@ class FlintrockCluster:
             user=user,
             identity_file=identity_file)
 
-    def generate_template_mapping(self, *, service: str) -> dict:
-        """
-        Generate a template mapping from a FlintrockCluster instance that we can use
-        to fill in template parameters.
-        """
-        root_dir = posixpath.join(self.storage_dirs.root, service)
-        ephemeral_dirs = ','.join(posixpath.join(path, service) for path in self.storage_dirs.ephemeral)
 
-        template_mapping = {
-            'master_ip': self.master_ip,
-            'master_host': self.master_host,
-            'slave_ips': '\n'.join(self.slave_ips),
-            'slave_hosts': '\n'.join(self.slave_hosts),
-            'root_dir': root_dir,
-            'ephemeral_dirs': ephemeral_dirs,
+def generate_template_mapping(
+    *,
+    cluster: FlintrockCluster,
+    # If we add additional services later on we may want to refactor
+    # this to take a list of services and dynamically pull the service
+    # name.
+    hadoop_version: str,
+    spark_version: str
+) -> dict:
+    """
+    Generate a template mapping from a FlintrockCluster instance that we can use
+    to fill in template parameters.
+    """
+    hadoop_root_dir = posixpath.join(cluster.storage_dirs.root, 'hadoop')
+    hadoop_ephemeral_dirs = ','.join(
+        posixpath.join(path, 'hadoop')
+        for path in cluster.storage_dirs.ephemeral
+    )
+    spark_root_dir = posixpath.join(cluster.storage_dirs.root, 'spark')
+    spark_ephemeral_dirs = ','.join(
+        posixpath.join(path, 'spark')
+        for path in cluster.storage_dirs.ephemeral
+    )
 
-            # If ephemeral storage is available, it replaces the root volume, which is
-            # typically persistent. We don't want to mix persistent and ephemeral
-            # storage since that causes problems after cluster stop/start; some volumes
-            # have leftover data, whereas others start fresh.
-            'root_ephemeral_dirs': ephemeral_dirs if ephemeral_dirs else root_dir,
-        }
+    template_mapping = {
+        'master_ip': cluster.master_ip,
+        'master_host': cluster.master_host,
+        'slave_ips': '\n'.join(cluster.slave_ips),
+        'slave_hosts': '\n'.join(cluster.slave_hosts),
 
-        return template_mapping
+        'hadoop_version': hadoop_version,
+        'hadoop_short_version': '.'.join(hadoop_version.split('.')[:2]),
+        'spark_version': spark_version,
+        'spark_short_version': '.'.join(spark_version.split('.')[:2]),
+
+        'hadoop_root_dir': hadoop_root_dir,
+        'hadoop_ephemeral_dirs': hadoop_ephemeral_dirs,
+        'spark_root_dir': spark_root_dir,
+        'spark_ephemeral_dirs': spark_ephemeral_dirs,
+
+        # If ephemeral storage is available, it replaces the root volume, which is
+        # typically persistent. We don't want to mix persistent and ephemeral
+        # storage since that causes problems after cluster stop/start; some volumes
+        # have leftover data, whereas others start fresh.
+        'hadoop_root_ephemeral_dirs': hadoop_ephemeral_dirs if hadoop_ephemeral_dirs else hadoop_root_dir,
+        'spark_root_ephemeral_dirs': spark_ephemeral_dirs if spark_ephemeral_dirs else spark_root_dir,
+    }
+
+    return template_mapping
+
+
+# TODO: Cache these files. (?) They are being read potentially tens or
+#       hundreds of times. Maybe it doesn't matter because the files
+#       are so small.
+def get_formatted_template(*, path: str, mapping: dict) -> str:
+    with open(path) as f:
+        formatted = f.read().format(**mapping)
+    return formatted
 
 
 def run_against_hosts(*, partial_func: functools.partial, hosts: list):

--- a/flintrock/flintrock.py
+++ b/flintrock/flintrock.py
@@ -187,13 +187,14 @@ def cli(cli_context, config, provider):
 @click.argument('cluster-name')
 @click.option('--num-slaves', type=click.IntRange(min=1), required=True)
 @click.option('--install-hdfs/--no-install-hdfs', default=False)
-@click.option('--hdfs-version')
+@click.option('--hdfs-version', default='2.7.3')
 @click.option('--hdfs-download-source',
               help="URL to download Hadoop from.",
               default='http://www.apache.org/dyn/closer.lua/hadoop/common/hadoop-{v}/hadoop-{v}.tar.gz?as_json',
               show_default=True)
 @click.option('--install-spark/--no-install-spark', default=True)
 @click.option('--spark-version',
+              default='2.1.0',
               help="Spark release version to install.")
 @click.option('--spark-download-source',
               help="URL to download a release of Spark from.",
@@ -314,7 +315,11 @@ def launch(
         services += [hdfs]
     if install_spark:
         if spark_version:
-            spark = Spark(version=spark_version, download_source=spark_download_source)
+            spark = Spark(
+                version=spark_version,
+                hadoop_version=hdfs_version,
+                download_source=spark_download_source,
+            )
         elif spark_git_commit:
             print(
                 "Warning: Building Spark takes a long time. "
@@ -324,7 +329,9 @@ def launch(
                 print("Building Spark at latest commit: {c}".format(c=spark_git_commit))
             spark = Spark(
                 git_commit=spark_git_commit,
-                git_repository=spark_git_repository)
+                git_repository=spark_git_repository,
+                hadoop_version=hdfs_version,
+            )
         services += [spark]
 
     if provider == 'ec2':

--- a/flintrock/services.py
+++ b/flintrock/services.py
@@ -9,7 +9,11 @@ import urllib.request
 import paramiko
 
 # Flintrock modules
-from .core import FlintrockCluster
+from .core import (
+    FlintrockCluster,
+    generate_template_mapping,
+    get_formatted_template,
+)
 from .ssh import ssh_check_output
 
 FROZEN = getattr(sys, 'frozen', False)
@@ -20,15 +24,6 @@ else:
     THIS_DIR = os.path.dirname(os.path.realpath(__file__))
 
 SCRIPTS_DIR = os.path.join(THIS_DIR, 'scripts')
-
-
-# TODO: Cache these files. (?) They are being read potentially tens or
-#       hundreds of times. Maybe it doesn't matter because the files
-#       are so small.
-def get_formatted_template(path: str, mapping: dict) -> str:
-    with open(path) as f:
-        formatted = f.read().format(**mapping)
-    return formatted
 
 
 class FlintrockService:
@@ -108,7 +103,7 @@ class FlintrockService:
 
 
 class HDFS(FlintrockService):
-    def __init__(self, version, download_source):
+    def __init__(self, *, version, download_source):
         self.version = version
         self.download_source = download_source
         self.manifest = {'version': version, 'download_source': download_source}
@@ -154,7 +149,8 @@ class HDFS(FlintrockService):
             'hadoop/conf/slaves',
             'hadoop/conf/hadoop-env.sh',
             'hadoop/conf/core-site.xml',
-            'hadoop/conf/hdfs-site.xml']
+            'hadoop/conf/hdfs-site.xml',
+        ]
 
         for template_path in template_paths:
             ssh_check_output(
@@ -165,7 +161,13 @@ class HDFS(FlintrockService):
                     f=shlex.quote(
                         get_formatted_template(
                             path=os.path.join(THIS_DIR, "templates", template_path),
-                            mapping=cluster.generate_template_mapping(service='hdfs'))),
+                            mapping=generate_template_mapping(
+                                cluster=cluster,
+                                hadoop_version=self.version,
+                                # Hadoop doesn't need to know what
+                                # Spark version we're using.
+                                spark_version='',
+                            ))),
                     p=shlex.quote(template_path)))
 
     # TODO: Convert this into start_master() and split master- or slave-specific
@@ -203,8 +205,15 @@ class HDFS(FlintrockService):
 
 
 class Spark(FlintrockService):
-    def __init__(self, version: str=None, download_source: str=None,
-                 git_commit: str=None, git_repository: str=None):
+    def __init__(
+        self,
+        *,
+        version: str=None,
+        hadoop_version: str,
+        download_source: str=None,
+        git_commit: str=None,
+        git_repository: str=None
+    ):
         # TODO: Convert these checks into something that throws a proper exception.
         #       Perhaps reuse logic from CLI.
         assert bool(version) ^ bool(git_commit)
@@ -212,12 +221,14 @@ class Spark(FlintrockService):
             assert git_repository
 
         self.version = version
+        self.hadoop_version = hadoop_version
         self.download_source = download_source
         self.git_commit = git_commit
         self.git_repository = git_repository
 
         self.manifest = {
             'version': version,
+            'hadoop_version': hadoop_version,
             'download_source': download_source,
             'git_commit': git_commit,
             'git_repository': git_repository}
@@ -261,13 +272,15 @@ class Spark(FlintrockService):
                         cd spark
                         git reset --hard {commit}
                         if [ -e "make-distribution.sh" ]; then
-                            ./make-distribution.sh -Phadoop-2.6
+                            ./make-distribution.sh -Phadoop-{hadoop_short_version}
                         else
-                            ./dev/make-distribution.sh -Phadoop-2.6
+                            ./dev/make-distribution.sh -Phadoop-{hadoop_short_version}
                         fi
                     """.format(
                         repo=shlex.quote(self.git_repository),
-                        commit=shlex.quote(self.git_commit)))
+                        commit=shlex.quote(self.git_commit),
+                        hadoop_short_version='.'.join(self.hadoop_version.split('.')[:2]),
+                    ))
             ssh_check_output(
                 client=ssh_client,
                 command="""
@@ -289,7 +302,9 @@ class Spark(FlintrockService):
             cluster: FlintrockCluster):
         template_paths = [
             'spark/conf/spark-env.sh',
-            'spark/conf/slaves']
+            'spark/conf/slaves',
+            'spark/conf/spark-defaults.conf',
+        ]
         for template_path in template_paths:
             ssh_check_output(
                 client=ssh_client,
@@ -299,7 +314,11 @@ class Spark(FlintrockService):
                     f=shlex.quote(
                         get_formatted_template(
                             path=os.path.join(THIS_DIR, "templates", template_path),
-                            mapping=cluster.generate_template_mapping(service='spark'))),
+                            mapping=generate_template_mapping(
+                                cluster=cluster,
+                                hadoop_version=self.hadoop_version,
+                                spark_version=self.version,
+                            ))),
                     p=shlex.quote(template_path)))
 
     # TODO: Convert this into start_master() and split master- or slave-specific

--- a/flintrock/templates/hadoop/conf/core-site.xml
+++ b/flintrock/templates/hadoop/conf/core-site.xml
@@ -4,7 +4,7 @@
 <configuration>
   <property>
     <name>hadoop.tmp.dir</name>
-    <value>{root_ephemeral_dirs}</value>
+    <value>{hadoop_root_ephemeral_dirs}</value>
   </property>
 
   <property>

--- a/flintrock/templates/hadoop/conf/hdfs-site.xml
+++ b/flintrock/templates/hadoop/conf/hdfs-site.xml
@@ -9,6 +9,6 @@
 
   <property>
     <name>dfs.datanode.data.dir</name>
-    <value>{root_ephemeral_dirs}</value>
+    <value>{hadoop_root_ephemeral_dirs}</value>
   </property>
 </configuration>

--- a/flintrock/templates/spark/conf/spark-defaults.conf
+++ b/flintrock/templates/spark/conf/spark-defaults.conf
@@ -1,0 +1,1 @@
+spark.jars.packages    org.apache.hadoop:hadoop-aws:{hadoop_version}

--- a/flintrock/templates/spark/conf/spark-env.sh
+++ b/flintrock/templates/spark/conf/spark-env.sh
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-export SPARK_LOCAL_DIRS="{root_ephemeral_dirs}"
+export SPARK_LOCAL_DIRS="{spark_root_ephemeral_dirs}"
 
 # Standalone cluster options
 export SPARK_EXECUTOR_INSTANCES="1"

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4,12 +4,38 @@ import tempfile
 import uuid
 from collections import OrderedDict
 
-# External modules
+# Flintrock
+from flintrock.core import FlintrockCluster, StorageDirs
+
+# External
 import pytest
 
 HADOOP_VERSION = '2.7.3'
-SPARK_VERSION = '2.0.2'
+SPARK_VERSION = '2.1.0'
 SPARK_GIT_COMMIT = '584354eaac02531c9584188b143367ba694b0c34'  # 2.0.2
+
+
+class Dummy():
+    pass
+
+
+@pytest.fixture(scope='session')
+def dummy_cluster():
+    storage_dirs = StorageDirs(
+        root='/media/root',
+        ephemeral=['/media/eph1', '/media/eph2'],
+        persistent=None,
+    )
+
+    cluster = Dummy()
+    cluster.name = 'test'
+    cluster.storage_dirs = storage_dirs
+    cluster.master_ip = '10.0.0.1'
+    cluster.master_host = 'master.hostname'
+    cluster.slave_ips = ['10.0.0.2']
+    cluster.slave_hosts = ['slave1.hostname']
+
+    return cluster
 
 
 def random_string():

--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -1,0 +1,29 @@
+import os
+
+# Flintrock
+from flintrock.core import (
+    generate_template_mapping,
+    get_formatted_template,
+)
+
+FLINTROCK_ROOT_DIR = (
+    os.path.dirname(
+        os.path.dirname(
+            os.path.realpath(__file__))))
+
+
+def test_templates(dummy_cluster):
+    template_dir = os.path.join(FLINTROCK_ROOT_DIR, 'flintrock', 'templates')
+    for (dirpath, dirnames, filenames) in os.walk(template_dir):
+        if filenames:
+            for filename in filenames:
+                template_path = os.path.join(dirpath, filename)
+                mapping = generate_template_mapping(
+                    cluster=dummy_cluster,
+                    hadoop_version='',
+                    spark_version='',
+                )
+                get_formatted_template(
+                    path=template_path,
+                    mapping=mapping,
+                )

